marlowe-actus/src/Language/Marlowe/ACTUS/Domain/Ops.hs

```diff
@@ -60,42 +57,71 @@ instance ActusOps Double where
     a / b       = a Prelude./ b
 
-instance DateOps LocalTime Double where
-    _lt a b = if a < b then _one else _zero
-
-instance YearFractionOps LocalTime Double where
+instance YearFractionOps Double where
     _y = yearFraction
 
-instance YearFractionOps (Value Observation) (Value Observation) where
-    _y DCC_A_360 _ _ _ = undefined
-    _y DCC_A_365 _ _ _ = undefined
-    _y _ _ _ _         = error "not available"
+instance ScheduleOps Double where
+    _ceiling = ceiling
+
+instance YearFractionOps (Value Observation) where
+    _y a b c d = Constant . toMarloweFixedPoint $ yearFraction a b c d
+      where
+        toMarloweFixedPoint = round <$> (fromIntegral marloweFixedPoint Prelude.*)
+
+instance ScheduleOps (Value Observation) where
+    _ceiling (Constant a) = a `div` marloweFixedPoint
+    -- ACTUS is implemented only for Fixed Schedules
+    -- that means schedules are known before the contract
+    -- is exectued, resp. the schedule do not depend on
+    -- riskfactors
+    _ceiling _            = error "Precondition: Fixed schedules"
 
 instance ActusOps (Value Observation) where
     _min a b = Cond (ValueLT a b) a b
     _max a b = Cond (ValueGT a b) a b
     _abs a = _max a (SubValue _zero a)
     _zero = Constant 0
     _one  = Constant marloweFixedPoint
-
-instance DateOps (Value Observation) (Value Observation) where
-    _lt a b = Cond (ValueLT a b) _one _zero
+    _fromInteger n = Constant $ n Prelude.* marloweFixedPoint
+    _negate a = NegValue a
 
 infixl 7  *, /
 infixl 6  +, -
 
 instance ActusNum (Value Observation) where
-    (+)                         = AddValue
-    (-)                         = SubValue
-    a * b                       = Scale (1 % marloweFixedPoint) $ MulValue a b
-    (Constant 0) / (Constant 0) = Constant 0 -- by convention in finance
-    (Constant x) / (Constant y) = Scale (marloweFixedPoint % 1) $ Constant $ div x y
-    x / (Constant y)            = Scale (marloweFixedPoint % y) x
-    x / y                       = DivValue x y
+  -- add
+  x + (Constant 0)            = x
+  (Constant 0) + y            = y
+  (Constant x) + (Constant y) = Constant (x Prelude.+ y)
+  x + y                       = AddValue x y
+
+  -- sub
+  x - (Constant 0)            = x
+  (Constant 0) - y            = NegValue y
+  (Constant x) - (Constant y) = Constant (x Prelude.- y)
+  x - y                       = SubValue x y
+
+  -- mul
+  (Constant 0) * _ = Constant 0
+  (Constant x) * (Constant y) = Constant (x Prelude.* y `div` marloweFixedPoint)
+  (Constant x) * y | x == marloweFixedPoint = y
+  (Constant x) * y | rem x marloweFixedPoint == 0 = MulValue (Constant (x `div` marloweFixedPoint)) y
+  x * (Constant y) | rem y marloweFixedPoint == 0 = MulValue x (Constant (y `div` marloweFixedPoint))
+  x * y = DivValue (MulValue x y) (Constant marloweFixedPoint)
+
+  -- div
+  (Constant 0) / _ = Constant 0
+  (Constant x) / (Constant y) = Constant $ x `div` (y Prelude.* marloweFixedPoint)
```

Comment:
:+1: but the nominator - well the brackets are wrong...

Reply:
In general I would prefer not have these special cases, for division for example `x/y = DivValue x y` is sufficient.
But this would generate huge Marlowe contracts that are no longer readable at all.

Reply:
Corrected as well the generic division case. The `marloweFixedPoint`  correction was also missing there.

Reply:
Thanks for spotting this!